iocore/net/SSLSessionCache.h

```diff
@@ -134,18 +166,19 @@ class SSLSessionBucket
 public:
   SSLSessionBucket();
   ~SSLSessionBucket();
-  void insertSession(const SSLSessionID &, SSL_SESSION *ctx, SSL *ssl);
-  bool getSession(const SSLSessionID &, SSL_SESSION **ctx, ssl_session_cache_exdata **data);
-  int getSessionBuffer(const SSLSessionID &, char *buffer, int &len);
-  void removeSession(const SSLSessionID &);
+  void insertSession(const SSLSessionID &sid, SSL_SESSION *sess, SSL *ssl);
+  bool getSession(const SSLSessionID &sid, SSL_SESSION **sess, ssl_session_cache_exdata **data);
+  int getSessionBuffer(const SSLSessionID &sid, char *buffer, int &len);
+  void removeSession(const SSLSessionID &sid);
 
 private:
   /* these method must be used while hold the lock */
   void print(const char *) const;
-  void removeOldestSession();
+  void removeOldestSession(const std::unique_lock<std::shared_mutex> &lock);
 
-  Ptr<ProxyMutex> mutex;
-  CountQueue<SSLSession> queue;
+  mutable std::shared_mutex mutex;
+  std::unordered_map<SSLSessionID, SSLSession *, SSLSessionIDHash> bucket_data;
+  std::map<ink_hrtime, SSLSession *> bucket_data_ts;
```

Comment:
I know `std::map` sorts items by key, but like I pointed out on my comment above, I don't think eviction order have to be that accurate. How about using `std::deque` instead?

Reply:
I was using a deque in the previous version, but then `SSLSessionBucket::removeSession` will need to go through the entire queue to delete an entry, which is why I switched to using a map.

Reply:
I see, then std::multimap would be another option. You can keep the number of entries as the same as the unordered_map. It still needs linear search but should be better than deque, and it doesn't require accurate time. So it's get_hrtime_updated vs linear search against entries that have the same timestamp. I'm not sure which is faster.